platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +83,131 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+            "'id' and/or 'expires_at' are missing in the GET /rate response",
+            ServerErrorException.class);
+      }
+    }
+
+    // sell_amount is present and positive number
+    if (!isPositiveNumber(Objects.requireNonNull(rate).getSellAmount())) {
+      logErrorAndThrow(
+          "'sell_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount is present and positive number
+    if (!isPositiveNumber(rate.getBuyAmount())) {
+      logErrorAndThrow(
+          "'buy_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is preset, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
+        logErrorAndThrow(
+            "'fee.total' is missing or not a positive number in the GET /rate response",
+            ServerErrorException.class);
+      }
+      // fee.asset is a valid asset
+      AssetInfo feeAsset = assetService.getAssetByName(fee.getAsset());
+      if (fee.getAsset() == null || feeAsset == null) {
+        logErrorAndThrow(
+            "'fee.asset' is missing or not a valid asset in the GET /rate response",
+            ServerErrorException.class);
+      }
+
+      if (fee.getAsset().equals(request.getSellAsset())) {
+        // when fee is in sell_asset,
+        // check that sell_amount is equal to price * buy_amount + (fee ?: 0)
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()))
+                .add(new BigDecimal(fee.getTotal()))
+                .setScale(sellAsset.getSignificantDecimals(), HALF_UP);
+
+        if (new BigDecimal(rate.getSellAmount()).compareTo(expected) != 0) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * buy_amount + (fee?:0) in the GET /rate response",
+              ServerErrorException.class);
+        }
+      } else {
+        // when fee is in buy_asset,
+        // check that sell_amount is equal to price * (buy_amount + (fee ?: 0))
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()).add(new BigDecimal(fee.getTotal())))
+                .setScale(sellAsset.getSignificantDecimals(), HALF_UP);
+        if (new BigDecimal(rate.getSellAmount()).compareTo(expected) != 0) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * (buy_amount + (fee ?: 0)) in the GET /rate response",
+              ServerErrorException.class);
+        }
+      }
+
+      if (fee.getDetails() != null) {
+        BigDecimal totalFee =
+            new BigDecimal(0)
+                .setScale(Objects.requireNonNull(feeAsset).getSignificantDecimals(), HALF_UP);
```

Comment:
I don't think you need to set scale here. 
1. When adding fees from the details it should already be counted on (scale would be <= significantDecimals and if it's not, the total check on line 194 would fail) 
2. compareTo doesn't care about the scale. 0 with scale 20 == 0 with scale 0 when used compareTo

Reply:
That's right, we don't need to set scale to validate the amount. If the response has floating errors caused by inconsistent decimals in the fee details, we should throw the error. 

I will remove the `setScale`.